### PR TITLE
Add Semantic Shift Benchmark

### PR DESCRIPTION
**Pull request by sgvaze**
# Task Description
We propose the Semantic Shift Benchmark suite (SSB) for open-set recognition and related tasks (e.g category discovery, out-of-distribution detection etc.). We define class splits from ImageNet-21K which are 'unseen' with respect to categories from ImageNet-1K.

Below, we show open-set classes (from ImageNet-21K) in dotted lines and their **closest counterparts** (from ImageNet-1K) in solid lines. The 'Hard' open-set classes have significantly more similar closed-set counterparts.

ssb_imagenet

## Dataset Creation
We mine open-set classes from [ImageNet-21K-P](https://github.com/Alibaba-MIIL/ImageNet21K), structuring the splits based on semantic distance in the WordNet hierarchy.

## Expected Insights/Relevance
The SSB is intended to isolate semantic novelty from other forms of distributional shifts. We structure open-set classes based on their semantic distance to the closed-set (the ImageNet-1K classes). Specifically, we provide 'seen' and 'unseen' class splits, where the 'unseen' classes are divided into 'Easy' and 'Hard'. Harder examples are more visually and semantically similar to the 'seen' classes.

## Evaluation Metrics
The unseen classes can be used for multiple downstream tasks (e.g category discovery and OoD detection). We evaluate open-set recognition performance with AUROC.

## Data
We use classes from the ImageNet-21K-P dataset, which can be downloaded and pre-processed following instructions [here](https://github.com/Alibaba-MIIL/ImageNet21K).

## License
ImageNet licence

### CODE REVIEW

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+#         super(ImageNetBase, self).__init__(root, transform)
+#
+#         self.uq_idxs = np.array(range(len(self)))
+#
+#     def __getitem__(self, item):
+#
+#         img, label = super().__getitem__(item)
+#         uq_idx = self.uq_idxs[item]
+#
+#         return img, label, uq_idx
```
Comment (stes) [inline shifthappens/tasks/ssb/imagenet_ssb.py:25]:
```suggestion
```

remove if not needed?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    hard_ssb_set = ImageNetSubsetDataset(root=os.path.join(imagenet21k_root, 'val'), classes_to_keep=hard_wnids,
+                                         transform=test_transform)
+
+    return easy_ssb_set, hard_ssb_set
+
+sh_benchmark.register_task(
+    name="SSB", relative_data_folder="imagenet21k_p", standalone=True
+)
+
+print('Hello')
```
Comment (stes) [inline shifthappens/tasks/ssb/imagenet_ssb.py:187]:
```suggestion
```

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        std = (0.229, 0.224, 0.225)
+
+        # test_transform = tv_transforms.Compose([
+        #     tv_transforms.Resize(256),
+        #     tv_transforms.CenterCrop(224),
+        #     tv_transforms.ToTensor(),
+        #     tv_transforms.Normalize(
+        #         mean=torch.tensor(mean),
+        #         std=torch.tensor(std))
+        # ])
```
Comment (stes) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:71]:
```suggestion
```

remove if not needed?

**Review by stes — CHANGES_REQUESTED** [review]
Thanks for the contribution!

A few minor comments to remove uncommented code --- also it might be useful to run the pre-commit hooks to format code etc., but @kotekjedi can also help with this and format the PR.

File: shifthappens/tasks/ssb/__init__.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+from pkg_resources import resource_filename
+from shifthappens import benchmark as sh_benchmark
+
+osr_split_path = resource_filename(__name__, 'imagenet_osr_splits_winter21.pkl')
+@sh_benchmark.register_task(
+    name="SSB", relative_data_folder="ssb", standalone=True
+)
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/__init__.py:7]:
```suggestion

```
Let's move the decorator to the `semantic_shift_benchmark.py` file.

File: shifthappens/tasks/ssb/imagenet_ssb.py (hunk 1 above)
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:25]:
```suggestion

```

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
+import torchvision
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:1]:
```suggestion
```
This import is never used.

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+import numpy as np
+import torch
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:3]:
```suggestion
```
This import is never used.

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import os
+import pickle
+from copy import deepcopy
+
+from torch.utils.data import Dataset
+from torchvision.datasets.folder import default_loader
+from shifthappens import benchmark as sh_benchmark
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:10]:
```suggestion
```
This import is never used.

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    hard_wnids = precomputed_info['hard_i21k_classes']
+
+    easy_ssb_set = ImageNetSubsetDataset(root=os.path.join(imagenet21k_root, 'val'), classes_to_keep=easy_wnids,
+                                         transform=test_transform)
+    hard_ssb_set = ImageNetSubsetDataset(root=os.path.join(imagenet21k_root, 'val'), classes_to_keep=hard_wnids,
+                                         transform=test_transform)
+
+    return easy_ssb_set, hard_ssb_set
+
+sh_benchmark.register_task(
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:183]: Why do we need second task registration here?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+    elif len(dataset2) > len(dataset1):
+
+        rand_idxs = np.random.choice(range(len(dataset2)), size=(len(dataset1, )))
+        subsample_dataset(dataset2, rand_idxs)
+
+    return dataset1, dataset2
+
+
+def get_imagenet_ssb_datasets(test_transform, imagenet21k_root, osr_split_path):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:165]: I would suggest passing .pkl here directly, smth like this:
```
def get_imagenet_ssb_datasets(test_transform, imagenet21k_root, osr_split):

    easy_wnids = osr_split['easy_i21k_classes']
    hard_wnids = osr_split['hard_i21k_classes']

```

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+The original dataset was collected by Dat Tran for the object detection task
+and can be found at https://github.com/datitran/raccoon_dataset.
+"""
+import dataclasses
+import os
+import pickle
+
+import numpy as np
+import torchvision.transforms as tv_transforms
+import torch
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:17]:
```suggestion
```
This import is never used.

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+import shifthappens.data.base as sh_data
+import shifthappens.data.torch as sh_data_torch
+import shifthappens.utils as sh_utils
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:21]:
```suggestion
```
This import is never used.

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+from shifthappens.models import base as sh_models
+from shifthappens.models.base import PredictionTargets
+from shifthappens.tasks.base import Task
+from shifthappens.tasks.metrics import Metric
+from shifthappens.tasks.mixins import OODScoreTaskMixin
+from shifthappens.tasks.task_result import TaskResult
+from shifthappens.tasks.utils import auroc_ood
+from shifthappens.tasks.utils import fpr_at_tpr
+from shifthappens.tasks.ssb.imagenet_ssb import get_imagenet_ssb_datasets
+from shifthappens.tasks.ssb import osr_split_path
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:32]: I would suggest removing this and add .pkl loading inside the class

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+from typing import Tuple
+
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:35]:
```suggestion
import requests

PKL_URL = "https://github.com/sgvaze/osr_closed_set_all_you_need/raw/main/data/open_set_splits/imagenet_osr_splits_winter21.pkl"


@sh_benchmark.register_task(name="SSB", relative_data_folder="ssb", standalone=True)
```

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+@dataclasses.dataclass
+class SSB(Task, OODScoreTaskMixin):
+
+    resource = (
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:39]: I would suggest removing this

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        None,
+        None,
+    )
+    dataset_out_easy: sh_data_torch.IndexedTorchDataset = None
+    dataset_out_hard: sh_data_torch.IndexedTorchDataset = None
+
+    max_batch_size: int = 256
+
+    def setup(self):
+
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:51]: Let's add downloading here

```
        pickle_path = os.path.join(self.data_root, "imagenet_osr_splits_winter21.pkl")
        if not os.path.exists(pickle_path):
            os.makedirs(self.data_root, exist_ok=True)
            pickle_file = requests.get(PKL_URL)
            open(pickle_path, "wb").write(pickle_file.content)

        self.osr_split = pickle.load(open(pickle_path, "rb"))

        imagenet_21k_root = os.path.join(self.data_root, "imagenet21k_resized_new")

```

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        imagenet_21k_root = os.path.join(self.data_root, imagenet_21k_dir_name)
+
+        # Ensure data is downloaded
+        assert (
+            self.assert_data_downloaded(imagenet_21k_root)
+        ), f'ImageNet-21K data not downloaded not found in {imagenet_21k_root} \n Please download and process according to:' \
+           ' https://github.com/Alibaba-MIIL/ImageNet21K/blob/main/dataset_preprocessing/processing_script.sh'
+
+        mean = (0.485, 0.456, 0.406)
+        std = (0.229, 0.224, 0.225)
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:62]:
```suggestion
```
This is never used.

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    max_batch_size: int = 256
+
+    def setup(self):
+
+        imagenet_21k_dir_name, ssb_split_path, url, md5 = self.resource
+        imagenet_21k_root = os.path.join(self.data_root, imagenet_21k_dir_name)
+
+        # Ensure data is downloaded
+        assert (
+            self.assert_data_downloaded(imagenet_21k_root)
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:57]: I would suggest passing here loaded .pkl:
` assert self.assert_data_downloaded(self.osr_split, imagenet_21k_root)
`

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+        test_transform = tv_transforms.Compose(
+            [
+                tv_transforms.ToTensor(),
+                tv_transforms.Lambda(lambda x: x.permute(1, 2, 0)),
+            ]
+        )
+
+        dataset_out_easy, dataset_out_hard = get_imagenet_ssb_datasets(imagenet21k_root=imagenet_21k_root,
+                                                                        osr_split_path=ssb_split_path,
+                                                                        test_transform=test_transform)
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:82]:
```
        dataset_out_easy, dataset_out_hard = get_imagenet_ssb_datasets(
            test_transform=test_transform,
            imagenet21k_root=imagenet_21k_root,
            osr_split=self.osr_split,
        )
```
I would suggest passing .pkl here.

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+        return result_easy, result_hard
+
+    @staticmethod
+    def assert_data_downloaded(imagenet21k_root):
+
+        try:
+
+            # Load splits
+            with open(osr_split_path, 'rb') as handle:
+                precomputed_info = pickle.load(handle)
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:145]:
```
    def assert_data_downloaded(osr_split, imagenet21k_root):
        try:
            easy_wnids = osr_split["easy_i21k_classes"]
            hard_wnids = osr_split["hard_i21k_classes"]
            all_wnids = easy_wnids.tolist() + hard_wnids

            dataset_folder = os.path.join(imagenet21k_root, "val")
```
Let's remove loading from here.

File: shifthappens/tasks/ssb/semantic_shift_benchmark.py [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+    def _evaluate(self, model: sh_models.Model) -> Tuple[TaskResult, TaskResult]:
+
+        dataloader_out_easy, dataloader_out_hard = self._prepare_dataloader()
+        result_easy = self._evaluate_single_split(model, dataloader_out_easy)
+        result_hard = self._evaluate_single_split(model, dataloader_out_hard)
+
+        return result_easy, result_hard
+
+    @staticmethod
+    def assert_data_downloaded(imagenet21k_root):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/semantic_shift_benchmark.py:139]: Should we maybe remove this function to `imagent_ssb.py` ?

**Comment by kotekjedi** [comment]
Should we maybe remove .png file from this PR?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        list2 = [None] * (len(list1) - len(list2)) + list2
+
+    else:
+
+        pass
+
+    return list1, list2
+
+
+def get_imagenet_osr_class_splits(imagenet21k_class_to_idx, precomputed_split_dir, osr_split='Easy'):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:62]: This function is never used - do we need it?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+
+    dataset.imgs = [x for i, x in enumerate(dataset.imgs) if i in idxs]
+    dataset.samples = [x for i, x in enumerate(dataset.samples) if i in idxs]
+    dataset.targets = np.array(dataset.targets)[idxs].tolist()
+    dataset.uq_idxs = dataset.uq_idxs[idxs]
+
+    return dataset
+
+
+def subsample_classes(dataset, include_classes=list(range(1000))):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:90]: This function is never used - do we need it?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+    for i, k in enumerate(include_classes):
+        target_xform_dict[k] = i
+
+    dataset = subsample_dataset(dataset, cls_idxs)
+    dataset.target_transform = lambda x: target_xform_dict[x]
+
+    return dataset
+
+
+def get_train_val_split(train_dataset, val_split=0.2):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:104]: This function is never used - do we need it?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+        val_idxs.extend(v_)
+
+    # Get training/validation datasets based on selected idxs
+    train_dataset = subsample_dataset(train_dataset, train_idxs)
+    val_dataset = subsample_dataset(val_dataset, val_idxs)
+
+    return train_dataset, val_dataset
+
+
+def get_equal_len_datasets(dataset1, dataset2):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:130]: This function is never used - do we need it?

**Comment by kotekjedi** [comment]
Hi Sagar! I refactored your code a bit, so it is now correctly integrated with our API. Could you please check the current version and say if everything is OK?

Also, it would be nice if you could add docstrings to `semantic_shift_benchmark.py `and remove unused functions from `imagenet_ssb.py` if they are unnecessary.

I would also like to clarify what version of the dataset task uses? Fall 11 Version or Winter 21 Version? I wrote a couple of asserts assuming that it is a Winter 21 Version, but if not, I will adapt the code. I think we should also note this in the docstring.

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+
+        self.root = root
+        self.classes = classes_to_keep
+        self.transform = transform
+
+        # List all samples
+        root_dirs = [os.path.join(root, c) for c in self.classes]
+
+        samples = ["image not found"] * len(root_dirs) * 50
+        print(len(samples))
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:23]:
```suggestion
```

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+        list2 = [None] * (len(list1) - len(list2)) + list2
+
+    else:
+
+        pass
+
+    return list1, list2
+
+
+def get_imagenet_osr_class_splits(
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:70]: This function is never used - do we need it?

File: shifthappens/tasks/ssb/imagenet_ssb.py [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+
+    dataset.imgs = [x for i, x in enumerate(dataset.imgs) if i in idxs]
+    dataset.samples = [x for i, x in enumerate(dataset.samples) if i in idxs]
+    dataset.targets = np.array(dataset.targets)[idxs].tolist()
+    dataset.uq_idxs = dataset.uq_idxs[idxs]
+
+    return dataset
+
+
+def subsample_classes(dataset, include_classes=range(1000)):
```
Comment (kotekjedi) [inline shifthappens/tasks/ssb/imagenet_ssb.py:98]: This function is never used - do we need it?

**Comment by sgvaze** [comment]
Hi @kotekjedi - thanks for your help! I have made the requested changes and pushed again :)

**Review by zimmerrol — APPROVED** [review]

**Review by zimmerrol — APPROVED** [review]

**Pull request merged**